### PR TITLE
Added support for relative symlinks in add_real_directory

### PR DESCRIPTION
**Pull request by luto**
Currently ignores `add_real_directory` for directories and flattens them for files. This PR adds support for relative dir and file symlinks by calling `create_symlink` within `add_real_directory`.

### CODE REVIEW

**Comment by luto** [comment]
> OSError: [Errno 63] File name too long: '/private/var/folders/nz/vv4_9tw56nv9k3tkvyszvwg80000gn/T/pip-req-build-M7VUbK/build/lib/pyfakefs/tests/fixtures/symlink_dir_relative/fixtures/symlink_dir_relative/fixtures/symlink_dir_relative/fixtures/symlink_dir_relative/fixtures/symlink_dir_relative/fixtures/symlink_dir_relative/fixtures/symlink_dir_relative/fixtures/symlink_dir_relative/fixtures/symlink_dir_relative/fixtures/symlink_dir_relative/fixtures/symlink_dir_relative/fixtures/symlink_dir_relative/fixtures/symlink_dir_relative/fixtures/symlink_dir_relative/fixtures/symlink_dir_relative/fixtures/symlink_dir_relative/fixtures/symlink_dir_relative/fixtures/symlink_dir_relative/fixtures/symlink_dir_relative/fixtures/symlink_dir_relative/fixtures/symlink_dir_relative/fixtures/symlink_dir_relative/fixtures/symlink_dir_relative/fixtures/symlink_dir_relative/fixtures/symlink_dir_relative/fixtures/symlink_dir_relative/fixtures/symlink_dir_relative/fixtures/symlink_dir_relative/fixtures/symlink_dir_relative/fixtures/symlink_dir_relative/all_tests_without_extra_packages.py'

ha :) I guess I'll have to create the symlinks dynamically during testing. That way I can also add support for absolute ones.

Before I do that, though, I'll wait for feedback :grin:

File: pyfakefs/tests/fixtures/symlink_dir_relative [hunk 1]
```diff
@@ -0,0 +1 @@
+..
```
[the remark this answers is not in the document]
Comment (luto) [reply pyfakefs/tests/fixtures/symlink_dir_relative]: This comment shows up on line 1 for me, so I'm not quite sure which code you're referring to. Can you move it, please?

**Comment by luto** [comment]
Thanks for the comments! I'll take a closer look on monday.

**Comment by mrbean-bremen** [comment]
Take your time! I was thinking about a new pypi release, as there is currently not much happening here, may be we will do that after this is finished (if you are not planning more changes, of course).

File: pyfakefs/tests/fake_filesystem_test.py [hunk 2]
```diff
@@ -1974,6 +1974,23 @@ def test_add_existing_real_directory_tree(self):
...
+                             'fixtures/symlink_dir_relative')))
+        self.assertTrue(
+            self.filesystem.exists(
+                os.path.join(self.root_path, 'pyfakefs', 'tests',
+                             'fixtures/symlink_dir_relative/all_tests.py')))
+        self.assertTrue(
+            self.filesystem.exists(
+                os.path.join(self.root_path, 'pyfakefs', 'tests',
+                             'fixtures/symlink_file_relative')))
+
```
[the remark this answers is not in the document]
Comment (luto) [reply pyfakefs/tests/fake_filesystem_test.py]: done

File: pyfakefs/fake_filesystem.py [hunk 3]
```diff
@@ -2476,8 +2476,29 @@ def add_real_directory(self, source_path, read_only=True, lazy_read=True,
...
+
+                        target = os.readlink(abs_fileEntry)
+
+                        if target.startswith('./') or target.startswith('../') or target in ('.', '..'):
+                            target = os.path.join(os.path.dirname(abs_fileEntry), target)
+                            target = os.path.normpath(target)
+                        else:
+                            continue
+
+                        if target.startswith(source_path + os.sep) or target == source_path:
```
[the remark this answers is not in the document]
Comment (luto) [reply pyfakefs/fake_filesystem.py]: Symlinks can now point to outside the added directory. So this is resolved, right?

File: pyfakefs/fake_filesystem.py [hunk 4]
```diff
@@ -2476,8 +2476,29 @@ def add_real_directory(self, source_path, read_only=True, lazy_read=True,
...
+                if self._is_link_supported():
+                    for fileEntry in os.listdir(base):
+                        abs_fileEntry = os.path.join(base, fileEntry)
+
+                        if not os.path.islink(abs_fileEntry):
+                            continue
+
+                        target = os.readlink(abs_fileEntry)
+
+                        if target.startswith('./') or target.startswith('../') or target in ('.', '..'):
```
[the remark this answers is not in the document]
Comment (luto) [reply pyfakefs/fake_filesystem.py]: That line is gone and has/needs no replacement :)

**Comment by luto** [comment]
I gave a code a good overhaul, making much simpler. There is now support for absolute symlinks, symlinks pointing to outside the added directory and `lazy_read=True`. I also accidentally introduced a new public API `add_real_symlink`, which is used with lazy read mode. Please let me know what you think!

**Comment by luto** [comment]
The windows tests mostly work now (:tada:), except for [`test_resolve_nonexisting_file_after_36` on py27](https://ci.appveyor.com/project/jmcgeheeiv/pyfakefs/builds/25641630/job/b3u6cobmp7y7ol2q). I'm not quite sure what is happening there and would be very happy about any pointers :sweat_smile:

File: pyfakefs/fake_filesystem.py [hunk 5]
```diff
@@ -2427,11 +2429,43 @@ def add_real_file(self, source_path, read_only=True, target_path=None):
                                fake_file.st_dev)
         return fake_file
 
+    def add_real_symlink(self, source_path, target_path=None):
+        """Create a at source_path (or target_path, if given).  It will point to
```
Comment (mrbean-bremen) [inline pyfakefs/fake_filesystem.py]: "Create a" - missing word here

File: pyfakefs/tests/fake_filesystem_test.py [hunk 6]
```diff
@@ -1974,6 +1975,111 @@ def test_add_existing_real_directory_tree(self):
...
 
+    @contextlib.contextmanager
+    def create_symlinks(self, symlinks):
+        for link in symlinks:
+            os.symlink(link[0], link[1])
+
+        yield
+
+        for link in symlinks:
+            os.unlink(link[1])
```
Comment (mrbean-bremen) [inline pyfakefs/tests/fake_filesystem_test.py]: Nice!

**Comment by luto** [comment]
> the minor typo in the doc string (missing "symlink")

fixed.

>the removal of the workaround in the unrelated failing test.

After looking at the pathlib2 issue as well as the git history of that test, I am not quite sure what the correct test would be. Additionally, I don't have a way to verify the behavior, since I'm lacking a windows system :/ Can you tackle that one, please?

> Thanks for the contribution!

Thanks for maintaining pyfakefs :grin:

**Comment by mrbean-bremen** [comment]
> Can you tackle that one, please?

Yes, of course!

**Comment by mrbean-bremen** [comment]
> After looking at the pathlib2 issue as well as the git history of that test, I am not quite sure what the correct test would be. 

Yes, turns out I misunderstood this too - the behavior is now reversed, but still seems to be wrong (ignores `strict=True`).

**Comment by luto** [comment]
Thanks! :)

**Pull request merged**